// @creatio-devkit/common/esm2020/lib/internal/services/element-registration-checker/element-registration-checker.mjs
import { camelCase, isEqual, upperFirst } from 'lodash';
/**
 * @internal
 */
export function checkTypeNameFormat(type, element) {
    const typeParts = type.split('.');
    if (typeParts.length < 2) {
        throw new Error(`Error when register ${type}. Type property should have format "VendorPrefix.${element}"`);
    }
}
/**
 * @public
 * @description For internal use only.
 */
export function ɵcheckTypeNameFormat(type, element) {
    return checkTypeNameFormat(type, element);
}
/**
 * @internal
 */
export function checkTypeNamePascalCase(type) {
    const [vendor, typeName] = type.split('.');
    const pascalTypeName = toPascalCase(typeName);
    if (!vendor || pascalTypeName !== typeName) {
        throw new Error(`Error when register ${type}. Type property should have Pascal case, e.g. ${vendor + '.' + pascalTypeName}. ` + 'The first letter of each word in a compound word should be capitalized');
    }
}
/**
 * @internal
 */
export function checkCrtTypeNamePascalCase(type) {
    const [vendor, typeName] = type.split('.');
    const pascalTypeName = toPascalCase(typeName);
    if (vendor.toLowerCase() === 'crt' && pascalTypeName !== typeName) {
        throw new Error(`Error when register ${type}. Type property should have Pascal case, e.g. ${vendor + '.' + pascalTypeName}. ` + 'The first letter of each word in a compound word should be capitalized');
    }
}
/**
 * @internal
 */
export function toPascalCase(source) {
    return upperFirst(camelCase(source));
}
/**
 * @internal
 */
export function checkIfTypeRegistered(type, item) {
    if (!item) {
        throw new Error(`Error when resolve ${type}. Type was not registered.`);
    }
}
/**
 * @internal
 */
export function checkOnlyOneRegistration(type, item, map) {
    if (map.has(type) && !isEqual(map.get(type), item)) {
        throw new Error(`Error when register ${type}. Element with this type has already exist`);
    }
}
