// @creatio/base/esm2020/lib/public/enums/data-value-type.enum.mjs
export var ɵDataValueType;
(function (ɵDataValueType) {
    ɵDataValueType[ɵDataValueType["Guid"] = 0] = "Guid";
    ɵDataValueType[ɵDataValueType["Text"] = 1] = "Text";
    ɵDataValueType[ɵDataValueType["Integer"] = 4] = "Integer";
    ɵDataValueType[ɵDataValueType["Float"] = 5] = "Float";
    ɵDataValueType[ɵDataValueType["Money"] = 6] = "Money";
    ɵDataValueType[ɵDataValueType["DateTime"] = 7] = "DateTime";
    ɵDataValueType[ɵDataValueType["Date"] = 8] = "Date";
    ɵDataValueType[ɵDataValueType["Time"] = 9] = "Time";
    ɵDataValueType[ɵDataValueType["Lookup"] = 10] = "Lookup";
    ɵDataValueType[ɵDataValueType["Enum"] = 11] = "Enum";
    ɵDataValueType[ɵDataValueType["Boolean"] = 12] = "Boolean";
    ɵDataValueType[ɵDataValueType["Blob"] = 13] = "Blob";
    ɵDataValueType[ɵDataValueType["Image"] = 14] = "Image";
    ɵDataValueType[ɵDataValueType["CUSTOM_OBJECT"] = 15] = "CUSTOM_OBJECT";
    ɵDataValueType[ɵDataValueType["IMAGELOOKUP"] = 16] = "IMAGELOOKUP";
    ɵDataValueType[ɵDataValueType["COLLECTION"] = 17] = "COLLECTION";
    ɵDataValueType[ɵDataValueType["Color"] = 18] = "Color";
    ɵDataValueType[ɵDataValueType["LOCALIZABLE_STRING"] = 19] = "LOCALIZABLE_STRING";
    ɵDataValueType[ɵDataValueType["ENTITY"] = 20] = "ENTITY";
    ɵDataValueType[ɵDataValueType["ENTITY_COLLECTION"] = 21] = "ENTITY_COLLECTION";
    ɵDataValueType[ɵDataValueType["ENTITY_COLUMN_MAPPING_COLLECTION"] = 22] = "ENTITY_COLUMN_MAPPING_COLLECTION";
    ɵDataValueType[ɵDataValueType["HASH_TEXT"] = 23] = "HASH_TEXT";
    ɵDataValueType[ɵDataValueType["SECURE_TEXT"] = 24] = "SECURE_TEXT";
    ɵDataValueType[ɵDataValueType["FILE"] = 25] = "FILE";
    ɵDataValueType[ɵDataValueType["MAPPING"] = 26] = "MAPPING";
    ɵDataValueType[ɵDataValueType["SHORT_TEXT"] = 27] = "SHORT_TEXT";
    ɵDataValueType[ɵDataValueType["MEDIUM_TEXT"] = 28] = "MEDIUM_TEXT";
    ɵDataValueType[ɵDataValueType["MAXSIZE_TEXT"] = 29] = "MAXSIZE_TEXT";
    ɵDataValueType[ɵDataValueType["LONG_TEXT"] = 30] = "LONG_TEXT";
    ɵDataValueType[ɵDataValueType["FLOAT1"] = 31] = "FLOAT1";
    ɵDataValueType[ɵDataValueType["FLOAT2"] = 32] = "FLOAT2";
    ɵDataValueType[ɵDataValueType["FLOAT3"] = 33] = "FLOAT3";
    ɵDataValueType[ɵDataValueType["FLOAT4"] = 34] = "FLOAT4";
    ɵDataValueType[ɵDataValueType["LOCALIZABLE_PARAMETER_VALUES_LIST"] = 35] = "LOCALIZABLE_PARAMETER_VALUES_LIST";
    ɵDataValueType[ɵDataValueType["METADATA_TEXT"] = 36] = "METADATA_TEXT";
    ɵDataValueType[ɵDataValueType["STAGE_INDICATOR"] = 37] = "STAGE_INDICATOR";
    ɵDataValueType[ɵDataValueType["OBJECT_LIST"] = 38] = "OBJECT_LIST";
    ɵDataValueType[ɵDataValueType["COMPOSITE_OBJECT_LIST"] = 39] = "COMPOSITE_OBJECT_LIST";
    ɵDataValueType[ɵDataValueType["FLOAT8"] = 40] = "FLOAT8";
    ɵDataValueType[ɵDataValueType["FILE_LOCATOR"] = 41] = "FILE_LOCATOR";
    ɵDataValueType[ɵDataValueType["PHONE_TEXT"] = 42] = "PHONE_TEXT";
    ɵDataValueType[ɵDataValueType["RICH_TEXT"] = 43] = "RICH_TEXT";
    ɵDataValueType[ɵDataValueType["WEB_TEXT"] = 44] = "WEB_TEXT";
    ɵDataValueType[ɵDataValueType["EMAIL_TEXT"] = 45] = "EMAIL_TEXT";
})(ɵDataValueType || (ɵDataValueType = {}));
