// @creatio/base/esm2020/lib/public/esq/data-model/expressions/arithmetic/arithmetic-expression.mjs
import { ɵExpressionType } from '../../enums';
import { ɵBaseExpression } from '../base-expression';
export class ɵArithmeticExpression extends ɵBaseExpression {
    /**
     * @internal
     * @dontChange
     */
    static get _instanceOfKeys() {
        return [...ɵBaseExpression._instanceOfKeys, this._instanceOfKey];
    }
    constructor(config) {
        super(config);
        this.expressionType = ɵExpressionType.ArithmeticOperation;
        this.arithmeticOperation = config?.arithmeticOperation;
        this.leftArithmeticOperand = config?.leftArithmeticOperand.clone();
        this.rightArithmeticOperand = config?.rightArithmeticOperand.clone();
    }
    /**
     * Parses expression from json.
     * @public
     */
    static fromJson(dto, filterParser, expressionParser) {
        const leftArithmeticOperand = expressionParser?.fromJson(dto['leftArithmeticOperand'], filterParser);
        const rightArithmeticOperand = expressionParser?.fromJson(dto['rightArithmeticOperand'], filterParser);
        const config = {
            leftArithmeticOperand,
            rightArithmeticOperand,
            arithmeticOperation: dto['arithmeticOperation'],
            isBlock: dto['isBlock'],
        };
        return new ɵArithmeticExpression(config);
    }
    clone() {
        return new ɵArithmeticExpression(this);
    }
}
/**
 * @internal
 * @dontChange
 */
ɵArithmeticExpression._instanceOfKey = 'devkit_ArithmeticExpression';
