// @creatio-devkit/common/esm2020/lib/public/services/handler-chain/handler-chain.service.mjs
import { InternalHandlerChain } from './internal-handler-chain';
const global = window;
/**
 * @public
 * @class HandlerChainService
 * @classdesc Service to design handlers in chain.
 */
export class HandlerChainService {
    get _chain() {
        return global.handlerChain;
    }
    set _chain(value) {
        global.handlerChain = value;
    }
    static get instance() {
        return HandlerChainService._instance || (HandlerChainService._instance = new HandlerChainService());
    }
    constructor() {
        if (!this._chain) {
            this._createGlobalHandlerChain();
        }
    }
    _createGlobalHandlerChain() {
        this._chain = new InternalHandlerChain();
    }
    /**
     * @internal
     */
    clearContextedHandlersCache(contextId) {
        return this._chain.clearContextedHandlersCache(contextId);
    }
    /**
     * @internal
     */
    register(handlerConfig) {
        return this._chain.handlerChain$.register(handlerConfig);
    }
    /**
     * @description Method to process passed request
     * @param request - {@link ɵBaseRequest}
     * @return `Promise<unknown>`
     */
    process(request) {
        return this._chain.handlerChain$.process(request);
    }
    /**
     * @internal
     * Subscribes to requests.
     * @param callback A handler function for each request triggered.
     * @param context (optional) A context in which the requests were triggered.
     * The `callback` will be called only for requests with specified context.
     * If not  specified, will be called for all requests.
     * @returns A function to cancel the subscription.
     */
    subscribe(callback, context) {
        return this._chain.handlerChain$.subscribe(callback, context);
    }
}
