// @creatio-devkit/common/esm2020/lib/internal/services/design-time-view-element-registry/design-time-view-element-registry.mjs
import { ɵBaseRegistry } from '../base-registry';
import { checkCrtTypeNamePascalCase, checkOnlyOneRegistration, checkTypeNameFormat, } from '../element-registration-checker';
/**
 * @internal
 */
export const designTimeViewElementRegistryKey = 'DesignTimeViewElementRegistry';
/**
 * @internal
 * @hostPublicApi available through window.creatio.registry.DesignTimeViewElementRegistry
 */
export class DesignTimeViewElementRegistry extends ɵBaseRegistry {
    static validate(config) {
        checkTypeNameFormat(config.type, 'DesignTimeViewElementRegistryItemTypeName');
        checkCrtTypeNamePascalCase(config.type);
        checkOnlyOneRegistration(config.type, config, this.infos);
    }
    static register(config) {
        this.infos.set(config.type, config);
    }
    static getByOverriddenType(type) {
        const item = Array.from(this.infos).find(([_, info]) => info.overrideType === type);
        if (item) {
            return item[1];
        }
        return null;
    }
    static get infos() {
        return this.store.typeInfos;
    }
}
DesignTimeViewElementRegistry.registryKey = designTimeViewElementRegistryKey;
