// @creatio/base/esm2020/lib/public/esq/data-model/enums/comparison-type.enum.mjs
/* eslint-disable @typescript-eslint/naming-convention */
export var ɵComparisonType;
(function (ɵComparisonType) {
    ɵComparisonType[ɵComparisonType["Between"] = 0] = "Between";
    ɵComparisonType[ɵComparisonType["Contain"] = 11] = "Contain";
    ɵComparisonType[ɵComparisonType["End_with"] = 13] = "End_with";
    ɵComparisonType[ɵComparisonType["Equal"] = 3] = "Equal";
    ɵComparisonType[ɵComparisonType["Exists"] = 15] = "Exists";
    ɵComparisonType[ɵComparisonType["Greater"] = 7] = "Greater";
    ɵComparisonType[ɵComparisonType["Greater_or_equal"] = 8] = "Greater_or_equal";
    ɵComparisonType[ɵComparisonType["Is_not_null"] = 2] = "Is_not_null";
    ɵComparisonType[ɵComparisonType["Is_null"] = 1] = "Is_null";
    ɵComparisonType[ɵComparisonType["Less"] = 5] = "Less";
    ɵComparisonType[ɵComparisonType["Less_or_equal"] = 6] = "Less_or_equal";
    ɵComparisonType[ɵComparisonType["Not_contain"] = 12] = "Not_contain";
    ɵComparisonType[ɵComparisonType["Not_end_with"] = 14] = "Not_end_with";
    ɵComparisonType[ɵComparisonType["Not_equal"] = 4] = "Not_equal";
    ɵComparisonType[ɵComparisonType["Not_exists"] = 16] = "Not_exists";
    ɵComparisonType[ɵComparisonType["Not_start_with"] = 10] = "Not_start_with";
    ɵComparisonType[ɵComparisonType["Start_with"] = 9] = "Start_with";
})(ɵComparisonType || (ɵComparisonType = {}));
