// @creatio/base/esm2020/lib/public/esq/data-model/enums/query-macros-type.enum.mjs
export var ɵQueryMacrosType;
(function (ɵQueryMacrosType) {
    /** Macro type not defined. */
    ɵQueryMacrosType[ɵQueryMacrosType["None"] = 0] = "None";
    /** Current user. */
    ɵQueryMacrosType[ɵQueryMacrosType["CurrentUser"] = 1] = "CurrentUser";
    /** Current user contact. */
    ɵQueryMacrosType[ɵQueryMacrosType["CurrentUserContact"] = 2] = "CurrentUserContact";
    /** Yesterday. */
    ɵQueryMacrosType[ɵQueryMacrosType["Yesterday"] = 3] = "Yesterday";
    /** Today. */
    ɵQueryMacrosType[ɵQueryMacrosType["Today"] = 4] = "Today";
    /** Tomorrow. */
    ɵQueryMacrosType[ɵQueryMacrosType["Tomorrow"] = 5] = "Tomorrow";
    /** Previous week. */
    ɵQueryMacrosType[ɵQueryMacrosType["PreviousWeek"] = 6] = "PreviousWeek";
    /** This week. */
    ɵQueryMacrosType[ɵQueryMacrosType["CurrentWeek"] = 7] = "CurrentWeek";
    /** Next week. */
    ɵQueryMacrosType[ɵQueryMacrosType["NextWeek"] = 8] = "NextWeek";
    /** Previous month. */
    ɵQueryMacrosType[ɵQueryMacrosType["PreviousMonth"] = 9] = "PreviousMonth";
    /** Current month. */
    ɵQueryMacrosType[ɵQueryMacrosType["CurrentMonth"] = 10] = "CurrentMonth";
    /** Next month. */
    ɵQueryMacrosType[ɵQueryMacrosType["NextMonth"] = 11] = "NextMonth";
    /** Previous quarter. */
    ɵQueryMacrosType[ɵQueryMacrosType["PreviousQuarter"] = 12] = "PreviousQuarter";
    /** Current quarter. */
    ɵQueryMacrosType[ɵQueryMacrosType["CurrentQuarter"] = 13] = "CurrentQuarter";
    /** Next quarter. */
    ɵQueryMacrosType[ɵQueryMacrosType["NextQuarter"] = 14] = "NextQuarter";
    /** Previous half of the year. */
    ɵQueryMacrosType[ɵQueryMacrosType["PreviousHalfYear"] = 15] = "PreviousHalfYear";
    /** Current half-year. */
    ɵQueryMacrosType[ɵQueryMacrosType["CurrentHalfYear"] = 16] = "CurrentHalfYear";
    /** Next half-year. */
    ɵQueryMacrosType[ɵQueryMacrosType["NextHalfYear"] = 17] = "NextHalfYear";
    /** Last year. */
    ɵQueryMacrosType[ɵQueryMacrosType["PreviousYear"] = 18] = "PreviousYear";
    /** This year. */
    ɵQueryMacrosType[ɵQueryMacrosType["CurrentYear"] = 19] = "CurrentYear";
    /** Previous hour. */
    ɵQueryMacrosType[ɵQueryMacrosType["PreviousHour"] = 20] = "PreviousHour";
    /** Current hour. */
    ɵQueryMacrosType[ɵQueryMacrosType["CurrentHour"] = 21] = "CurrentHour";
    /** Next hour. */
    ɵQueryMacrosType[ɵQueryMacrosType["NextHour"] = 22] = "NextHour";
    /** Next year. */
    ɵQueryMacrosType[ɵQueryMacrosType["NextYear"] = 23] = "NextYear";
    /** Next N Days. */
    ɵQueryMacrosType[ɵQueryMacrosType["NextNDays"] = 24] = "NextNDays";
    /** Previous N days. */
    ɵQueryMacrosType[ɵQueryMacrosType["PreviousNDays"] = 25] = "PreviousNDays";
    /** Next N hours. */
    ɵQueryMacrosType[ɵQueryMacrosType["NextNHours"] = 26] = "NextNHours";
    /** Previous N hours. */
    ɵQueryMacrosType[ɵQueryMacrosType["PreviousNHours"] = 27] = "PreviousNHours";
    /** Primary column. */
    ɵQueryMacrosType[ɵQueryMacrosType["PrimaryColumn"] = 34] = "PrimaryColumn";
    /** Primary displayed column. */
    ɵQueryMacrosType[ɵQueryMacrosType["PrimaryDisplayColumn"] = 35] = "PrimaryDisplayColumn";
    /** Primary image display column. */
    ɵQueryMacrosType[ɵQueryMacrosType["PrimaryImageColumn"] = 36] = "PrimaryImageColumn";
    /** Anniversary today. */
    ɵQueryMacrosType[ɵQueryMacrosType["DayOfYearToday"] = 37] = "DayOfYearToday";
    /** Anniversary on the date computed as today plus days offset. */
    ɵQueryMacrosType[ɵQueryMacrosType["DayOfYearTodayPlusDaysOffset"] = 38] = "DayOfYearTodayPlusDaysOffset";
    /** Anniversary on the next several days. */
    ɵQueryMacrosType[ɵQueryMacrosType["NextNDaysOfYear"] = 39] = "NextNDaysOfYear";
    /** Anniversary on the previous several days. */
    ɵQueryMacrosType[ɵQueryMacrosType["PreviousNDaysOfYear"] = 40] = "PreviousNDaysOfYear";
    /** Primary color display column. */
    ɵQueryMacrosType[ɵQueryMacrosType["PrimaryColorColumn"] = 41] = "PrimaryColorColumn";
})(ɵQueryMacrosType || (ɵQueryMacrosType = {}));
