// @creatio-devkit/common/esm2020/lib/internal/services/view-element-registry/view-element-registry.mjs
import { ɵBaseRegistry } from '../base-registry';
import { checkOnlyOneRegistration, checkTypeNameFormat } from '../element-registration-checker';
/**
 * @internal
 */
export const viewElementRegistryKey = 'ViewElementRegistry';
/**
 * @internal
 * @hostPublicApi available through window.creatio.registry.ValidatorRegistry
 */
export class ViewElementRegistry extends ɵBaseRegistry {
    static validate(config) {
        checkTypeNameFormat(config.type, 'ViewElementTypeName');
        checkOnlyOneRegistration(config.type, config, ViewElementRegistry.viewElements);
    }
    static get viewElements() {
        return this.store.typeInfos;
    }
    static register(config) {
        this.viewElements.set(config.type, config);
    }
    static getFormControlRelatesAttribute(type) {
        const config = this.store.typeInfos.get(type);
        return (config && 'formControlConfig' in config && config.formControlConfig?.relatesTo) || null;
    }
}
ViewElementRegistry.registryKey = viewElementRegistryKey;
