// @creatio-devkit/common/esm2020/lib/internal/services/validator-registry/validator-registry.mjs
import { ɵBaseRegistry } from '../base-registry';
import { checkTypeNameFormat } from '../element-registration-checker';
/**
 * @internal
 */
export class ValidatorMapValue {
}
/**
 * @internal
 */
export const validatorRegistryKey = 'ValidatorRegistry';
/**
 * @internal
 * @hostPublicApi available through window.creatio.registry.ValidatorRegistry
 */
export class ValidatorRegistry extends ɵBaseRegistry {
    static _has(configType, type) {
        return ValidatorRegistry.items.has(configType) && ValidatorRegistry.items.get(configType)?.typeRef !== type;
    }
    static validate(config) {
        checkTypeNameFormat(config.type, 'ValidatorTypeName');
        if (ValidatorRegistry._has(config.type, config.typeRef)) {
            throw new Error(`Error when register ${config.type}. Element with this type has already exist`);
        }
    }
    static get items() {
        return this.store.typeInfos;
    }
    static register(config) {
        this.items.set(config.type, config);
    }
}
ValidatorRegistry.registryKey = validatorRegistryKey;
