// @creatio/base/esm2020/lib/public/esq/data-model/expressions/sub-queries/sub-query-expression.mjs
import { ɵExpressionType } from '../../enums';
import { ɵBaseExpression } from '../base-expression';
export class ɵSubQueryExpression extends ɵBaseExpression {
    /**
     * @internal
     * @dontChange
     */
    static get _instanceOfKeys() {
        return [...ɵBaseExpression._instanceOfKeys, this._instanceOfKey];
    }
    constructor(config) {
        super(config);
        this.expressionType = ɵExpressionType.SubQuery;
        this.columnPath = config?.columnPath;
        const subFilter = config?.subFilters;
        this.subFilters = typeof subFilter?.clone === 'function' ? subFilter.clone() : subFilter;
        this.subOrderDirection = config.subOrderDirection;
        this.subOrderColumn = config.subOrderColumn;
    }
    /**
     * Parses expression from json.
     * @public
     */
    static fromJson(dto, filterParser) {
        const expression = new ɵSubQueryExpression({
            columnPath: dto['columnPath'],
            subFilters: filterParser.fromJson(dto['subFilters']),
            subOrderDirection: dto['subOrderDirection'],
            subOrderColumn: dto['subOrderColumn'],
        });
        return expression;
    }
    clone() {
        return new ɵSubQueryExpression(this);
    }
    toJson() {
        const result = super.toJson();
        if (this.subFilters) {
            result['subFilters'] = this.serializeItem(this.subFilters);
        }
        return result;
    }
}
/**
 * @internal
 * @dontChange
 */
ɵSubQueryExpression._instanceOfKey = 'devkit_SubQueryExpression';
