// @creatio/base/esm2020/lib/public/esq/data-model/expressions/functions/aggregation-function-expression.mjs
import { ɵArgumentFunctionExpression } from './argument-function-expression';
import { ɵFunctionType } from '../../enums';
export class ɵAggregationFunctionExpression extends ɵArgumentFunctionExpression {
    /**
     * @internal
     * @dontChange
     */
    static get _instanceOfKeys() {
        return [...ɵArgumentFunctionExpression._instanceOfKeys, this._instanceOfKey];
    }
    constructor(config) {
        super(config);
        this.functionType = ɵFunctionType.Aggregation;
        this.aggregationType = config?.aggregationType;
        this.aggregationEvalType = config?.aggregationEvalType;
    }
    /**
     * Parses expression from json.
     * @public
     */
    static fromJson(dto, filterParser, expressionParser) {
        const functionArgument = expressionParser?.fromJson(dto['functionArgument'], filterParser);
        const aggregationType = dto['aggregationType'];
        const aggregationEvalType = dto['aggregationEvalType'];
        return new ɵAggregationFunctionExpression({
            functionArgument,
            aggregationType,
            aggregationEvalType,
            isBlock: dto['isBlock'],
        });
    }
    clone() {
        return new ɵAggregationFunctionExpression(this);
    }
}
/**
 * @internal
 * @dontChange
 */
ɵAggregationFunctionExpression._instanceOfKey = 'devkit_AggregationFunctionExpression';
